src/GafferScene/Instancer.cpp

```diff
@@ -2356,3 +2525,319 @@ void Instancer::PrototypeScope::setPrototype( const EngineData *engine, const Sc
+			&renderer, &hasAttributes
+		]
+		( const PrototypeCacheGetterKey &key, size_t &cost, const IECore::Canceller *canceller ) -> ConstPrototypePtr
+		{
+			cost = 1;
+			return new Prototype(
+				prototypesPlug, prototypeRoot, sampleTimes, outerCapsuleHash, renderOptions,
+				key.context, renderer,
+				// If we don't have instance attributes, we can prepare renderer attributes ahead of time
+				!hasAttributes
+			);
+		},
+		pointIndicesForPrototype.size() // Never evict, even if prototypes are all unique
+	);
+
+	// We've found problems with performance when running too many iterations in parallel, which appear
+	// to be related with hitting AiNode too hard in parallel ( perhaps related to threads spread between
+	// separate processors ). To partially solve this, we set the grain size so that we shouldn't use more
+	// than 32 threads, which appears to help some in testing.
+	size_t grainSize = std::max( (size_t)1, pointIndicesForPrototype.size() / 32 );
+	task_group_context taskGroupContext( task_group_context::isolated );
+	tbb::parallel_for( tbb::blocked_range<size_t>( 0, pointIndicesForPrototype.size(), grainSize ), [&]( const tbb::blocked_range<size_t> &r )
+		{
+			vector<M44f> pointTransforms( sampleTimes.size() );
+			std::string name;
+			IECoreScenePreview::Renderer::AttributesInterfacePtr attribsStorage;
+
+			Context::EditableScope prototypeScope( scope.context() );
+
+			for( size_t idx = r.begin(); idx != r.end(); ++idx )
+			{
+				int pointIndex = pointIndicesForPrototype[idx];
+
+				const Prototype *proto;
+				if( constantPrototype )
+				{
+					proto = constantPrototype.get();
+				}
+				else
+				{
+					// The prototype depends on the context, so we need to find the prototype context for
+					// this instance.
+
+
+					// We find the capsules using the engine at shutter open, but the time used to construct the capsules
+					// must be the on-frame time, since the capsules will add their own shutter ( and we also handle
+					// the shutter ourselves for transform matrices )
+					//
+					// For most context variables, we are overwriting them for each prototype anyway, so
+					// we can reuse the context. But timeOffset is relative, so it's important that we reset the
+					// time before we do setPrototypeContextVariables for the next element. ( Should this be more
+					// general instead of assuming that frame is the only variable for which offsetMode may be set? )
+					prototypeScope.setFrame( onFrameTime );
+
+					engines[0]->setPrototypeContextVariables( pointIndex, prototypeScope );
+
+					proto = prototypeCache.get( PrototypeCacheGetterKey( prototypeScope.context() ) ).get();
+				}
+
+				if( !proto->m_object.size() )
+				{
+					// No object to render. This could happen if the protype didn't meet the
+					// RenderOptions::purposeIncluded test.
+					continue;
+				}
+
+				IECoreScenePreview::Renderer::AttributesInterface *attribs;
+				int instanceId = engines[0]->instanceId( pointIndex );
```

Comment:
`instanceId` isn't used in the "get the attributes" section, so it could be moved to L2807 where is is actually first used.